File: python/invariance/batcher.py
from __future__ import annotations

import asyncio
from typing import Callable

from .errors import InvarianceError
from .http_client import HttpClient
from .types import Receipt


class Batcher:
    def __init__(
        self,
        *,
        http: HttpClient,
        flush_interval_ms: int = 5000,
        max_batch_size: int = 50,
        max_queue_size: int = 1000,
        on_error: Callable[[InvarianceError], None] | None = None,
    ) -> None:
        self._http = http
        self._flush_interval_ms = flush_interval_ms
        self._max_batch_size = max_batch_size
        self._max_queue_size = max_queue_size
        self._on_error = on_error
        self._queue: list[Receipt] = []
        self._pending: list[asyncio.Task[None]] = []
        self._timer_task: asyncio.Task[None] | None = None
        self._start_timer()

    def _start_timer(self) -> None:
        try:
            loop = asyncio.get_running_loop()
            self._timer_task = loop.create_task(self._timer_loop())
        except RuntimeError:
            # No running event loop yet — timer will not auto-flush
            self._timer_task = None

    async def _timer_loop(self) -> None:
        try:
            while True:
                await asyncio.sleep(self._flush_interval_ms / 1000.0)
                if self._queue:
                    await self.flush()
        except asyncio.CancelledError:
            pass

    def enqueue(self, receipt: Receipt) -> None:
        if len(self._queue) >= self._max_queue_size:
            overflow = len(self._queue) - self._max_queue_size + 1
            self._queue = self._queue[overflow:]
            if self._on_error:
                self._on_error(
                    InvarianceError(
                        "QUEUE_OVERFLOW",
                        f"Queue overflow: dropped {overflow} receipts",
                    )
                )
        self._queue.append(receipt)

        if len(self._queue) >= self._max_batch_size:
            try:
                loop = asyncio.get_running_loop()
                task = loop.create_task(self.flush())
                self._pending.append(task)
                task.add_done_callback(lambda t: self._pending.remove(t) if t in self._pending else None)
            except RuntimeError:
                pass

    async def flush(self) -> None:
        if not self._queue:
            return

        batch = self._queue[: self._max_batch_size]
        self._queue = self._queue[self._max_batch_size :]
        await self._send_batch(batch)

    async def shutdown(self) -> None:
        if self._timer_task:
            self._timer_task.cancel()
            try:
                await self._timer_task
            except asyncio.CancelledError:
                pass
            self._timer_task = None

        if self._queue:
            await self.flush()

        if self._pending:
            await asyncio.gather(*self._pending, return_exceptions=True)

    async def _send_batch(self, batch: list[Receipt]) -> None:
        try:
            await self._http.post("/v1/receipts", {"receipts": batch})
        except InvarianceError as error:
            if error.status_code and error.status_code >= 500:
                self._queue = batch + self._queue
            if self._on_error:
                self._on_error(error)
        except Exception as exc:
            error = InvarianceError("FLUSH_FAILED", str(exc))
            if self._on_error:
                self._on_error(error)

File: packages/typescript/src/batcher.ts
import type { HttpClient } from './http.js';
import type { InvarianceError } from './errors.js';
import type { Receipt } from './types/receipt.js';

export interface BatcherConfig {
  http: HttpClient;
  flushIntervalMs?: number;
  maxBatchSize?: number;
  maxQueueSize?: number;
  onError?: (error: InvarianceError) => void;
}

export class Batcher {
  private queue: Receipt[] = [];
  private timer: ReturnType<typeof setInterval> | null = null;
  private pending: Promise<void>[] = [];
  private http: HttpClient;
  private flushIntervalMs: number;
  private maxBatchSize: number;
  private maxQueueSize: number;
  private onError?: (error: InvarianceError) => void;

  constructor(config: BatcherConfig) {
    this.http = config.http;
    this.flushIntervalMs = config.flushIntervalMs ?? 5000;
    this.maxBatchSize = config.maxBatchSize ?? 50;
    this.maxQueueSize = config.maxQueueSize ?? 1000;
    this.onError = config.onError;

    this.timer = setInterval(() => {
      if (this.queue.length > 0) this.flush();
    }, this.flushIntervalMs);

    // Don't prevent Node from exiting
    if (this.timer && typeof this.timer === 'object' && 'unref' in this.timer) {
      this.timer.unref();
    }
  }

  enqueue(receipt: Receipt): void {
    if (this.queue.length >= this.maxQueueSize) {
      // Drop oldest receipts to make room
      const overflow = this.queue.length - this.maxQueueSize + 1;
      this.queue.splice(0, overflow);
      this.onError?.({
        name: 'InvarianceError',
        message: `Queue overflow: dropped ${overflow} receipts`,
        code: 'QUEUE_OVERFLOW',
      } as InvarianceError);
    }
    this.queue.push(receipt);

    if (this.queue.length >= this.maxBatchSize) {
      this.flush();
    }
  }

  flush(): Promise<void> {
    if (this.queue.length === 0) return Promise.resolve();

    const batch = this.queue.splice(0, this.maxBatchSize);
    const promise = this.sendBatch(batch);
    this.pending.push(promise);
    promise.finally(() => {
      const idx = this.pending.indexOf(promise);
      if (idx !== -1) this.pending.splice(idx, 1);
    });
    return promise;
  }

  async shutdown(): Promise<void> {
    if (this.timer) {
      clearInterval(this.timer);
      this.timer = null;
    }

    // Flush remaining
    if (this.queue.length > 0) {
      await this.flush();
    }

    // Await all pending
    await Promise.allSettled(this.pending);
  }

  private async sendBatch(batch: Receipt[]): Promise<void> {
    try {
      await this.http.post('/v1/receipts', { receipts: batch });
    } catch (err) {
      const error = err as InvarianceError;
      // Re-queue on 5xx (server error)
      if (error.statusCode && error.statusCode >= 500) {
        this.queue.unshift(...batch);
      }
      this.onError?.(error);
    }
  }
}
